src/librustc/util/fs.rs

```diff
@@ -56,14 +56,49 @@ pub fn fix_windows_verbatim_for_gcc(p: &Path) -> PathBuf {
     }
 }
 
+pub enum LinkOrCopy {
+    Link,
+    Copy
+}
+
 /// Copy `p` into `q`, preferring to use hard-linking if possible. If
 /// `q` already exists, it is removed first.
-pub fn link_or_copy<P: AsRef<Path>, Q: AsRef<Path>>(p: P, q: Q) -> io::Result<()> {
+/// The result indicates which of the to operations has been performed.
+pub fn link_or_copy<P: AsRef<Path>, Q: AsRef<Path>>(p: P, q: Q) -> io::Result<LinkOrCopy> {
     let p = p.as_ref();
     let q = q.as_ref();
     if q.exists() {
         try!(fs::remove_file(&q));
     }
-    fs::hard_link(p, q)
-        .or_else(|_| fs::copy(p, q).map(|_| ()))
+
+    match fs::hard_link(p, q) {
```

Comment:
AFAICT simple copy is not reflink on ReFS, sadly.

There’s [FSCTL](https://msdn.microsoft.com/en-us/library/windows/desktop/mt590823(v=vs.85).aspx) in a Windows Server 2016+ which seems like it should behave like the reflink equivalent for btrfs.

That being said, it also seems like ReFSv1 (which is what you get by default) does not in fact support/expose that operation, so I guess there’s nothing we can do here.